ACMILabs/acmi-api
# #19 Add search to the API

Resolves #19

Add `/search/` to the API using elasticsearch.

### Acceptance Criteria
- [x] Add Elasticsearch to the development environment
- [x] Add search functionality to `/search/`
- [x] Index Works after importing them
- [x] Add production [Elastic Cloud](https://cloud.elastic.co/deployments/d88e6c3ec25f429aa7a3c49757a82c15) instance
- [x] Index `works` on Elastic Cloud
- [x] Match `/search/` results to `/works/` results
- [x] Add credentials to `acmi-devops` (https://github.com/ACMILabs/acmi-devops/issues/425)

### Relevant design files
* None

### Testing instructions
1. Start Docker: `cd development` and `docker-compose up --build`
1. See that search fails with a reasonable message: http://localhost:8081/search/?query=xos
1. Set `UPDATE_SEARCH=true` and `DEBUG=true` to your `config.env`
1. Stop Docker and restart it to index the search: `docker-compose up`
1. See that the search now returns results: http://localhost:8081/search/?query=xos
1. Try filtering by title: http://localhost:8081/search/?query=xos&field=title
1. Try result page sizes: http://localhost:8081/search/?query=xos&size=2&page=3
1. Now test out the production search: `UPDATE_SEARCH=false` and `DEBUG=false` to your `config.env`
1. Get `ELASTICSEARCH_CLOUD_ID` and `ELASTICSEARCH_API_KEY` for your `config.env` from 1Password: [Azure Elastic Cloud ID](https://start.1password.com/open/i?a=257GF32Z2VBSNCOB6TKMERSAAM&v=w5ej5dxmymsn5lff2egssofwuq&i=nooihb234jhsr5ye2ctsnj5lfa&h=acmiteam.1password.com) and [Azure Elastic Cloud API key](https://start.1password.com/open/i?a=257GF32Z2VBSNCOB6TKMERSAAM&v=w5ej5dxmymsn5lff2egssofwuq&i=n2gt7x47rrbf5kc2xoplnwjfqy&h=acmiteam.1password.com)
1. Start just the API: `cd development` and `docker-compose -f docker-compose-base.yml up`
1. See that the production search works: http://localhost:8081/search/?query=xos&size=2&page=3

### DoD
For requester to complete:
~- [ ] Deployment / migration instruction have been updated if required~
